src/Adapters/Secondary/DataInMemory/Repositories/ProductRepositoryInMemory.ts

```diff
@@ -0,0 +1,114 @@
+import IProductRepository from '@Application/Ports/Secondary/IProductRepository'
+import Product from '@Application/domain/Entities/Product'
+import { Either, Left, Right } from '@Shared/util/either'
+import { randomUUID } from 'crypto'
+import NotFoundException from '../../../../Application/domain/Exceptions/NotFoundException'
+
+type ProductType = {
+    id: Product['id']
+    name: Product['name']
+    category: Product['category']
+    price: Product['price']
+    description: Product['description']
+}
+
+export default class ProductRepositoryInMemory implements IProductRepository {
+    private list: ProductType[]
+
+    constructor() {
+        this.list = []
+    }
+
+    async create(product: Product): Promise<Either<Error, string>> {
+        const id = randomUUID()
+
+        this.list.push({
+            id,
+            name: product.getName(),
+            category: product.getCategory(),
+            price: product.getPrice(),
+            description: product.getDescription(),
+        })
+
+        return Right<string>('Product has been created')
+    }
+
+    async update(product: Product): Promise<Either<Error, string>> {
+        const existingProductIndex = this.list.findIndex(
+            (productFind) => productFind.id === product.getId()
+        )
+
+        if (existingProductIndex <= -1) {
+            return Left<Error>(new NotFoundException('Product: ' + product.getId()))
+        }
+
+        this.list[existingProductIndex] = {
+            id: this.list[existingProductIndex].id,
+            name: product.getName(),
+            category: product.getCategory(),
+            price: product.getPrice(),
+            description: product.getDescription(),
+        }
+
+        return Right('Product has been updated')
+    }
+
+    async delete(id: string): Promise<Either<Error, string>> {
+        const existingProductIndex = this.list.findIndex(
+            (productFind) => productFind.id === id
+        )
+
+        if (existingProductIndex <= -1) {
+            Left<Error>(new NotFoundException('Product: ' + id))
+        }
+
+        this.list.splice(existingProductIndex, 1)
+        return Right<string>("Product removed")
+    }
+
+    async findByCategory(category: string): Promise<Either<Error, Product[]>> {
+        let products = undefined
+        const productSaved = this.list.filter(
+            (productFind) => productFind.category === category
+        )
+
+        if (productSaved) {
```

Comment:
**suggestion** (non-blocking): early return

We can improve this part doing the early return, this way we can be more objective and avoid callback hell